libnmstate/nm/context.py

```diff
@@ -62,12 +100,98 @@ def _del_client(self):
                 timeout_source = GLib.timeout_source_new(50)
                 try:
                     timeout_source.set_callback(lambda x: is_timeout.append(1))
-                    timeout_source.attach(self._context)
+                    timeout_source.attach(self.context)
                     while not is_done and not is_timeout:
-                        self._context.iteration(True)
+                        self.context.iteration(True)
                 finally:
                     timeout_source.destroy()
-
             if not is_done:
                 logging.error("BUG: NM.Client is not cleaned")
             self._context = None
+
+    def _del_timeout(self):
+        if self._timeout_source:
+            self._timeout_source.destroy()
+            self._timeout_source = None
+
+    def register_async(self, action, fast=False):
+        """
+        Register action(string) to wait list.
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            # TODO: No need to wait all finish, a single finished async is
+            #       sufficient.
+            self.wait_all_finish()
+
+        logging.debug(f"Async action: {action} started")
+        queue.add(action)
+
+    def finish_async(self, action, suppress_log=False):
```

Comment:
This `suppress_log` is not nice. Can we hide it in the object state? Then it will be decided in one place (hopefully) and not passed on per call.

The naming here is not exactly corresponding to the counterpart: register-finish.
It is also not finishing anything, it just unregister the action.
But I will just be glad to see same name groups, e.g.: register/unregister, start/finish, add/remove, etc... At the usage it will also be much clearer that they represent the opposite of the other.

Reply:
Can you elaborate on `hide it in the object state`?
That argument is only introduced when reapply failed and fallback to normal activation. I don't
want people get confused about the log saying reapply finished.

I intend to hide how the async actions been done by `nm/context.py`. Register here means context will run async action in their own preferred time and finished means don't wait this async action anymore as it's done.

Reply:
I thought that you can set this flag at the object level when created, but per your response it seems that it always is suppose to log it except specific scenarios. So I guess there is nothing we can do.

Regarding the naming, I still think that register-finish counterparts are confusing, but I will not insist here.